test/e2e/external_oidc_test.go

```diff
@@ -0,0 +1,468 @@
+}
+
+func TestExternalOIDCWithKeycloak(t *testing.T) {
+	testCtx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	tc, err := newTestClient(t)
+	require.NoError(t, err)
+
+	// ===============================
+	// Check ExternalOIDC feature gate
+	// ===============================
+
+	featureGates, err := tc.configClient.ConfigV1().FeatureGates().Get(testCtx, "cluster", metav1.GetOptions{})
+	require.NoError(tc.t, err)
+
+	if len(featureGates.Status.FeatureGates) != 1 {
+		// fail test if there are multiple feature gate versions (i.e. ongoing upgrade)
+		tc.t.Fatalf("multiple feature gate versions detected")
+	} else {
+		for _, fgDisabled := range featureGates.Status.FeatureGates[0].Disabled {
+			if fgDisabled.Name == featureGateExternalOIDC {
+				tc.t.Skipf("feature gate '%s' disabled", featureGateExternalOIDC)
+			}
+		}
+	}
+
+	// ==============
+	// Setup Keycloak
+	// ==============
+
+	var kcClient *test.KeycloakClient
+	if keycloakURL := os.Getenv("E2E_KEYCLOAK_URL"); len(keycloakURL) > 0 {
+		t.Logf("will use existing keycloak deployment at URL: %s", keycloakURL)
+		kcClient = tc.setupKeycloakClient(testCtx, keycloakURL)
+
+	} else {
+		t.Logf("no existing keycloak deployment found; will create new")
+		var cleanups []func()
+		kcClient, cleanups = tc.setupExternalOIDCWithKeycloak(testCtx)
+		defer test.IDPCleanupWrapper(func() {
+			for _, c := range cleanups {
+				c()
+			}
+		})()
+		t.Logf("keycloak Admin URL: %s", kcClient.AdminURL())
+	}
+
+	kcAdminClient, err := kcClient.GetClientByClientID(oidcClientId)
+	require.NoError(t, err)
+	require.NotEmpty(t, kcAdminClient)
+
+	group := names.SimpleNameGenerator.GenerateName("e2e-keycloak-group-")
+	err = kcClient.CreateGroup(group)
+	require.NoError(t, err)
+
+	user := names.SimpleNameGenerator.GenerateName("e2e-keycloak-user-")
+	email := fmt.Sprintf("%s@test.dev", user)
+	password := "password"
+	err = kcClient.CreateUser(user, email, password, []string{group})
+	require.NoError(t, err)
+
+	httpClient := &http.Client{Transport: &http.Transport{
+		TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
+	}}
+
+	formData := url.Values{
+		"grant_type": []string{"password"},
+		"client_id":  []string{oidcClientId},
+		"scope":      []string{"openid"},
+		"username":   []string{user},
+		"password":   []string{password},
+	}
+
+	resp, err := httpClient.PostForm(kcClient.TokenURL(), formData)
+	require.NoError(t, err)
+	defer resp.Body.Close()
+	require.Equal(t, http.StatusOK, resp.StatusCode)
+
+	data, err := io.ReadAll(resp.Body)
+	require.NoError(t, err)
+	require.NotEmpty(t, data)
+
+	var authResponse oidcAuthResponse
+	err = json.Unmarshal(data, &authResponse)
+	require.NoError(t, err)
+	require.NotEmpty(t, authResponse.AccessToken)
+	require.NotEmpty(t, authResponse.IdToken)
+
+	// ==========================================
+	// Test authentication via the kube-apiserver
+	// ==========================================
```

Comment:
Was there a particular reason to avoid using a Kubernetes client for this part?

Reply:
Great idea, fixed in d89947d01557fc41b98fbbf551ea8a8ca0e196bc